sdk/core/azure-core-amqp/src/amqp/connection.cpp

```diff
@@ -556,13 +589,74 @@ namespace Azure { namespace Core { namespace Amqp { namespace _detail {
     return false;
   }
 
-  std::string ConnectionImpl::GetSecurityToken(
+  // Ensure that we have a token for the provided audience.
+  // If we don't, authenticate the audience with the service using the provided session.
+  // Note that the granularity of
+  Credentials::AccessToken ConnectionImpl::AuthenticateAudience(
+      std::shared_ptr<SessionImpl> session,
       std::string const& audience,
-      Azure::Core::Context const& context) const
+      Azure::Core::Context const& context)
   {
     if (GetCredential())
     {
-      if (m_tokenStore.find(audience) == m_tokenStore.end())
+      std::string audienceUrl = audience;
+      if (m_options.EnableTrace)
+      {
+        Log::Stream(Logger::Level::Verbose) << "Authenticate connection for audience " << audience;
+      }
+      // If the audience looks like a URL for AMQP, AMQPS, or SB, we can use the URL as provided.
+      if ((audience.find("amqps://") != 0) && (audience.find("amqp://") != 0)
+          && (audience.find("sb://") != 0))
+      {
+        audienceUrl = "amqps://" + GetHost();
+        // The provided audience may begin with a /, if not, we need to add the separator.
+        if (audience.front() != '/')
+        {
+          audienceUrl += "/";
+        }
+        audienceUrl += audience;
+        if (m_options.EnableTrace)
+        {
+          Log::Stream(Logger::Level::Verbose)
+              << "Initial audience is not URL, using " << audienceUrl;
+        }
+      }
+
+      std::unique_lock<std::mutex> lock(m_tokenMutex);
```

Comment:
Why would we want to use a std::shared_lock when we always want exclusive access to the lock?

Reply:
If I read thi correctly, this case, it does not seem that we are modifying the token store - we search, and if found, we return it, otherwise we throw. If so, multiple threads can read the thing at the same time by using shared_lock, and places that modify the collection or its items would use unique_lock.